### ISSUE TITLE
[BUG] [azure-storage-common]: StorageImplUtils initialization fails with java.security.AccessControlException when run under SecurityManager

### ISSUE THREAD

**Issue by reta** [Azure/azure-sdk-for-java#37464]
**Describe the bug**
The `StorageImplUtils` initialization fails with `java.security.AccessControlException` when run under `SecurityManager`.

***Exception or Stack Trace***
```
        java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "shutdownHooks")                                                                                                                                                                                                                                           
            at java.base/java.security.AccessControlContext.checkPermission(AccessControlContext.java:488)                                                                 
            at java.base/java.security.AccessController.checkPermission(AccessController.java:1071)                                                                        
            at java.base/java.lang.SecurityManager.checkPermission(SecurityManager.java:411)                                                                               
            at java.base/java.lang.Runtime.addShutdownHook(Runtime.java:248)                                                                                                                                                                                                                                                                          
            at com.azure.storage.common.implementation.StorageImplUtils.registerShutdownHook(StorageImplUtils.java:489)                                                    
            at com.azure.storage.common.implementation.StorageImplUtils.getThreadPoolWithShutdownHook(StorageImplUtils.java:471)                                           
            at com.azure.storage.common.implementation.StorageImplUtils.<clinit>(StorageImplUtils.java:98)                                                                 
            ... 13 more  
```

**To Reproduce**
```
System.setSecurityManager(...);
var threadPool = StorageImplUtils.THREAD_POOL; // trigger StorageImplUtils initialization
```

***Code Snippet***
```
Runtime.getRuntime().addShutdownHook(hook);
```

**Expected behavior**
The code should work normally under `SecurityManager`.

**Screenshots**

**Setup (please complete the following information):**
 - OS: any
 - IDE: any
 - Library/Libraries: azure-storage-common:12.23.1
 - Java version: any
 - App Server/Environment: any
 - Frameworks: any

If you suspect a dependency version mismatch (e.g. you see `NoClassDefFoundError`, `NoSuchMethodError` or similar), please check out [Troubleshoot dependency version conflict article](https://aka.ms/azsdk/java/dependency/troubleshoot) first. If it doesn't provide solution for the problem, please provide:
- verbose dependency tree (`mvn dependency:tree -Dverbose`)
- exception message, full stack trace, and any available logs

**Additional context**

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [X] Bug Description Added
- [X] Repro Steps Added
- [ ] Setup information Added

**Comment by alzimmermsft** [comment]
Thanks for filing this issue and submitting a PR @reta.

@ibrahimrabab, @ibrandes could you take a look at the PR.

**Comment by ibrahimrabab** [comment]
Hi @reta 
I'm trying to reproduce this locally but not seeing any specific errors being thrown, can you provide me a full code snippet that reproduces the error? Thanks!

**Comment by reta** [comment]
@ibrahimrabab please find the reproducer below:

```
package com.azure.repr;

import com.azure.storage.common.implementation.StorageImplUtils;

public class AppMain {

    public static void main(String[] args) {
        // In general we use policy file but this setup simplifies the reproducer
        Policy.setPolicy(new Policy() {
    		@Override
    		public PermissionCollection getPermissions(CodeSource codesource) {
    			final Permissions permissions = new Permissions();
    			permissions.add(new java.lang.RuntimePermission("accessClassInPackage.*"));
    			permissions.add(new java.lang.RuntimePermission("getenv.*"));
    			permissions.add(new java.util.PropertyPermission("*", "read"));
    			return permissions;
    		}
    	});
    	System.setSecurityManager(new SecurityManager());
        System.out.println(StorageImplUtils.THREAD_POOL);
    }

}
```

```
<project xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<groupId>com.azure</groupId>
	<artifactId>repr</artifactId>
	<version>0.0.1-SNAPSHOT</version>

	<properties>
		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
		<maven.compiler.source>11</maven.compiler.source>
		<maven.compiler.target>11</maven.compiler.target>
	</properties>

	<build>
		<plugins>
			<plugin><!-- Use latest to build recent Java applications -->
				<artifactId>maven-compiler-plugin</artifactId>
				<version>3.8.1</version>
			</plugin>
		</plugins>
	</build>

	<dependencies>
		<dependency>
			<groupId>com.azure</groupId>
			<artifactId>azure-storage-common</artifactId>
			<version>12.23.1</version>
		</dependency>
	</dependencies>
</project>
```

Running the AppMain ends with:

```
Exception in thread "main" java.lang.ExceptionInInitializerError
	at com.azure.repr.AppMain.main(AppMain.java:24)
Caused by: java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "shutdownHooks")
	at java.base/java.security.AccessControlContext.checkPermission(AccessControlContext.java:472)
	at java.base/java.security.AccessController.checkPermission(AccessController.java:897)
	at java.base/java.lang.SecurityManager.checkPermission(SecurityManager.java:322)
	at java.base/java.lang.Runtime.addShutdownHook(Runtime.java:214)
	at com.azure.storage.common.implementation.StorageImplUtils.registerShutdownHook(StorageImplUtils.java:489)
	at com.azure.storage.common.implementation.StorageImplUtils.getThreadPoolWithShutdownHook(StorageImplUtils.java:471)
	at com.azure.storage.common.implementation.StorageImplUtils.<clinit>(StorageImplUtils.java:98)
	... 1 more
```

**Comment by alzimmermsft** [comment]
@reta are the permissions missing a switch on the codebase to allow the azure-sdk-for-java code base to have `shutdownHooks` permissions?

@ibrahimrabab you could use the testing logic for JacksonAdapter to validate this: 

https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core/src/test/java/com/azure/core/util/serializer/JacksonAdapterSecurityIT.java

https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core/src/test/resources/JacksonAdapterSecurityPolicies/basic-permissions.policy

https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core/src/test/resources/JacksonAdapterSecurityPolicies/access-helper-succeeds.policy

The basic permissions doesn't give the azure-sdk-for-java code base permission to run the privileged action while the access helper succeeds does.

**Comment by reta** [comment]
> @reta are the permissions missing a switch on the codebase to allow the azure-sdk-for-java code base to have `shutdownHooks` permissions?

Thanks @alzimmermsft , there are two parts:
1) missed permission (it has to be added indeed, not added in the example to demonstrate the issue)
2) the AccessController.doPrivileged to make a call with caller's permissions

The caller chain we have is a bit more complicated, I could 100% reproduce it but the sample project would get a bit more complicated (since there are more components involved).

```
Caused by: java.lang.ExceptionInInitializerError: Exception java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "shutdownHooks") [in thread "TEST-AzureBlobContainerRetriesTests.testReadRangeBlobWithRetries-seed#[56A343CDB838E480]"]
	at java.base/java.security.AccessControlContext.checkPermission(AccessControlContext.java:488)
	at java.base/java.security.AccessController.checkPermission(AccessController.java:1071)
	at java.base/java.lang.SecurityManager.checkPermission(SecurityManager.java:411)
	at java.base/java.lang.Runtime.addShutdownHook(Runtime.java:248)
	at com.azure.storage.common.implementation.StorageImplUtils.registerShutdownHook(StorageImplUtils.java:489)
	at com.azure.storage.common.implementation.StorageImplUtils.getThreadPoolWithShutdownHook(StorageImplUtils.java:471)
	at com.azure.storage.common.implementation.StorageImplUtils.<clinit>(StorageImplUtils.java:98)
	at com.azure.storage.common.policy.RequestRetryOptions.<init>(RequestRetryOptions.java:98)
	at com.azure.storage.common.policy.RequestRetryOptions.<init>(RequestRetryOptions.java:64)
	at org.opensearch.repositories.azure.AzureBlobContainerRetriesTests$1.createRetryPolicy(AzureBlobContainerRetriesTests.java:176)
	at org.opensearch.repositories.azure.AzureStorageService.applyLocationMode(AzureStorageService.java:228)
	at org.opensearch.repositories.azure.AzureStorageService.buildClient(AzureStorageService.java:208)
	at org.opensearch.repositories.azure.AzureStorageService.lambda$client$1(AzureStorageService.java:152)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at org.opensearch.repositories.azure.AzureStorageService.client(AzureStorageService.java:150)
	at org.opensearch.repositories.azure.AzureBlobStore.client(AzureBlobStore.java:396)
	at org.opensearch.repositories.azure.AzureBlobStore.getInputStream(AzureBlobStore.java:276)
	at org.opensearch.repositories.azure.AzureBlobContainer.openInputStream(AzureBlobContainer.java:103)
	at org.opensearch.repositories.azure.AzureBlobContainer.readBlob(AzureBlobContainer.java:121)
	at org.opensearch.repositories.azure.AzureBlobContainerRetriesTests.testReadRangeBlobWithRetries(AzureBlobContainerRetriesTests.java:304)
	... 38 more

```

**Comment by reta** [comment]
@alzimmermsft @ibrahimrabab folks, any chances to get this issue addressed, the pull request is already out? (we are stuck with older releases) thank you

**Comment by alzimmermsft** [comment]
This should be fixed soon by #38729, using a similar pattern that was used when this issue came up with `JacksonAdapter`

**Comment by alzimmermsft** [comment]
HI @reta, this will be available in the `azure-storage-common` `12.25.0` release that should happen either later today or tomorrow. To enable `AccessController.doPrivileged` you'll need to set `AZURE_ENABLE_SHUTDOWN_HOOK_WITH_PRIVILEGE` to true

**Comment by reta** [comment]
> HI @reta, this will be available in the `azure-storage-common` `12.25.0` release that should happen either later today or tomorrow. To enable `AccessController.doPrivileged` you'll need to set `AZURE_ENABLE_SHUTDOWN_HOOK_WITH_PRIVILEGE` to true

Awesome, thanks a lot for the update @alzimmermsft !
